### PR TITLE
Raven db 14123

### PR DESCRIPTION
**Pull request by maximburyak**
Fixed several subscriptions related issues
https://issues.hibernatingrhinos.com/issue/RavenDB-14123
https://issues.hibernatingrhinos.com/issue/RavenDB-14180
https://issues.hibernatingrhinos.com/issue/RavenDB-14181
https://issues.hibernatingrhinos.com/issue/RavenDB-14182

### CODE REVIEW

File: src/Raven.Server/Documents/TcpHandlers/SubscriptionConnection.cs [hunk 1]
```diff
@@ -142,24 +165,24 @@ public async Task InitAsync()
                 }
                 catch (TimeoutException)
                 {
-                    if (timeout == TimeSpan.Zero && _logger.IsInfoEnabled)
+                    if (timeout == TimeSpan.FromMilliseconds(16) && _logger.IsInfoEnabled)
```
Comment (arekpalinski) [inline src/Raven.Server/Documents/TcpHandlers/SubscriptionConnection.cs]: Why exactly 16 milliseconds? Is it correct?

Comment (maximburyak) [reply src/Raven.Server/Documents/TcpHandlers/SubscriptionConnection.cs]: Yes, it's timout's initial value. There is no scenario where it can be TimeSpan.Zero

File: src/Raven.Server/Documents/Handlers/SubscriptionsHandler.cs [hunk 2]
```diff
@@ -380,7 +380,9 @@ private static DynamicJsonValue GetSubscriptionConnectionDJV(SubscriptionConnect
                 [nameof(SubscriptionConnection.ClientUri)] = x.ClientUri,
                 [nameof(SubscriptionConnection.Strategy)] = x.Strategy,
                 [nameof(SubscriptionConnection.Stats)] = GetConnectionStatsDJV(x.Stats),
-                [nameof(SubscriptionConnection.ConnectionException)] = x.ConnectionException?.Message
+                [nameof(SubscriptionConnection.ConnectionException)] = x.ConnectionException?.Message,
+                ["TCPConnectionStats"] = x.TcpConnection.GetConnectionStats(),
```
Comment (ayende) [inline src/Raven.Server/Documents/Handlers/SubscriptionsHandler.cs]: `nameof` ?

File: src/Raven.Server/Documents/TcpHandlers/SubscriptionConnection.cs [hunk 3]
```diff
@@ -66,6 +67,26 @@ public class SubscriptionConnection : IDisposable
         public long SubscriptionId { get; set; }
         public SubscriptionOpeningStrategy Strategy => _options.Strategy;
 
+        private ConcurrentQueue<string> _statusDescription = new ConcurrentQueue<string>();
+
+        public string[] GetRecentStatusDescriptions()
+        {
+            return _statusDescription?.ToArray()??Array.Empty<string>();
```
Comment (ayende) [inline src/Raven.Server/Documents/TcpHandlers/SubscriptionConnection.cs]: when can this be null?

File: src/Raven.Server/Documents/TcpHandlers/SubscriptionConnection.cs (hunk 1 above)
Comment (ayende) [reply src/Raven.Server/Documents/TcpHandlers/SubscriptionConnection.cs]: Let's make this a constant value, instead.

File: src/Raven.Server/Documents/TcpHandlers/SubscriptionConnection.cs [hunk 4]
```diff
@@ -515,7 +545,7 @@ private async Task ProcessSubscriptionAsync()
                                 subscriptionChangeVectorBeforeCurrentBatch = _lastChangeVector ?? SubscriptionState.ChangeVectorForNextBatchStartingPoint;
 
                                 if (sendingCurrentBatchStopwatch.ElapsedMilliseconds > 1000)
-                                    await SendHeartBeat();
+                                    await SendHeartBeat("Didnt' find any documents to send and more then 1000ms passed");
```
Comment (ayende) [inline src/Raven.Server/Documents/TcpHandlers/SubscriptionConnection.cs]:
```suggestion
                                    await SendHeartBeat("Didn't find any documents to send and more then 1000ms passed");
```

File: src/Raven.Server/Documents/TcpHandlers/SubscriptionConnection.cs [hunk 5]
```diff
@@ -695,16 +727,9 @@ private async Task<bool> TrySendingBatchToClient(DocumentsOperationContext docsC
                         if (_buffer.Length > Constants.Size.Megabyte ||
                             sendingCurrentBatchStopwatch.ElapsedMilliseconds > 1000)
                         {
-                            if (docsToFlush > 0)
```
Comment (ayende) [inline src/Raven.Server/Documents/TcpHandlers/SubscriptionConnection.cs:698]: If there is nothing to flush, nothing will be sent the user.
Why did you remove the `SendHeartBeat` ?

File: src/Raven.Server/Documents/TcpHandlers/SubscriptionConnection.cs (hunk 3 above)
Comment (maximburyak) [reply src/Raven.Server/Documents/TcpHandlers/SubscriptionConnection.cs]: After SubscriptionConnection being disposed

File: src/Raven.Server/Documents/TcpHandlers/SubscriptionConnection.cs (hunk 5 above)
Comment (maximburyak) [reply src/Raven.Server/Documents/TcpHandlers/SubscriptionConnection.cs:698]: There is no way docsToFlush will be 0, look few lines above

File: src/Raven.Server/Documents/Handlers/SubscriptionsHandler.cs (hunk 2 above)
Comment (maximburyak) [reply src/Raven.Server/Documents/Handlers/SubscriptionsHandler.cs]: I think that in that case, especially because it's a debug endpoint, it's more important to maintain clearer function name then to maintain "strongly typed" structure

File: src/Raven.Server/Documents/TcpHandlers/SubscriptionConnection.cs (hunk 1 above)
Comment (maximburyak) [reply src/Raven.Server/Documents/TcpHandlers/SubscriptionConnection.cs]: ok

File: src/Raven.Server/Documents/TcpHandlers/SubscriptionConnection.cs (hunk 4 above)
Comment (maximburyak) [reply src/Raven.Server/Documents/TcpHandlers/SubscriptionConnection.cs]: ok

File: src/Raven.Server/Documents/Handlers/SubscriptionsHandler.cs (hunk 2 above)
Comment (ppekrol) [reply src/Raven.Server/Documents/Handlers/SubscriptionsHandler.cs]: For consistency with other endpoints, this should be `Tcp` instead of `TCP`

File: src/Raven.Server/Documents/Handlers/SubscriptionsHandler.cs [hunk 6]
```diff
@@ -380,7 +380,9 @@ private static DynamicJsonValue GetSubscriptionConnectionDJV(SubscriptionConnect
...
+                ["StatusSubscriptions"] = new DynamicJsonArray(x.GetRecentStatusDescriptions())
```
Comment (ppekrol) [inline src/Raven.Server/Documents/Handlers/SubscriptionsHandler.cs]: The method here is `GetRecentStatusDescriptions` but the property is `StatusSubscriptions`. Is this valid?

File: src/Raven.Server/Documents/Subscriptions/SubscriptionConnectionState.cs [hunk 7]
```diff
@@ -18,9 +18,10 @@ public class SubscriptionConnectionState : IDisposable
         private readonly SubscriptionStorage _storage;
         internal readonly AsyncManualResetEvent ConnectionInUse = new AsyncManualResetEvent();
 
-        public SubscriptionConnectionState(long subscriptionId, SubscriptionStorage storage)
+        public SubscriptionConnectionState(long subscriptionId, string subscriptionName, SubscriptionStorage storage)
         {
             _subscriptionId = subscriptionId;
+            SubscriptionName = subscriptionName;
```
Comment (ppekrol) [inline src/Raven.Server/Documents/Subscriptions/SubscriptionConnectionState.cs]: check if not null?

File: src/Raven.Server/Documents/TcpHandlers/SubscriptionConnection.cs [hunk 8]
```diff
@@ -66,6 +68,26 @@ public class SubscriptionConnection : IDisposable
         public long SubscriptionId { get; set; }
         public SubscriptionOpeningStrategy Strategy => _options.Strategy;
 
+        private ConcurrentQueue<string> _statusDescription = new ConcurrentQueue<string>();
```
Comment (ppekrol) [inline src/Raven.Server/Documents/TcpHandlers/SubscriptionConnection.cs]: typo: `statusDescription`

File: src/Raven.Server/Documents/TcpHandlers/SubscriptionConnection.cs [hunk 9]
```diff
@@ -142,24 +166,24 @@ public async Task InitAsync()
...
                 {
-                    if (timeout == TimeSpan.Zero && _logger.IsInfoEnabled)
+                    if (timeout == InitialConnectionTimeout && _logger.IsInfoEnabled)
                     {
                         _logger.Info(
                             $"Subscription Id {SubscriptionId} from IP {TcpConnection.TcpClient.Client.RemoteEndPoint} starts to wait until previous connection from {_connectionState.Connection?.TcpConnection.TcpClient.Client.RemoteEndPoint} is released");
                     }
                     timeout = TimeSpan.FromMilliseconds(Math.Max(250, (long)_options.TimeToWaitBeforeConnectionRetry.TotalMilliseconds / 2));
-                    await SendHeartBeat();
+                    await SendHeartBeat($"Client from IP Waiting {TcpConnection.TcpClient.Client.RemoteEndPoint} waiting for subscription that is serving IP {_connectionState.Connection?.TcpConnection.TcpClient.Client.RemoteEndPoint} to be released");
```
Comment (ppekrol) [inline src/Raven.Server/Documents/TcpHandlers/SubscriptionConnection.cs:166]: is this first `Waiting` needed?

**Comment by ppekrol** [comment]
Beside those small remarks LGTM

**Comment by ayende** [comment]
@ppekrol Agreed, LTGM

File: src/Raven.Server/Documents/Handlers/SubscriptionsHandler.cs (hunk 6 above)
Comment (maximburyak) [reply src/Raven.Server/Documents/Handlers/SubscriptionsHandler.cs]: NP, I'll change this property to be a public readonly and create a copy of it upon consumption

File: src/Raven.Server/Documents/Subscriptions/SubscriptionConnectionState.cs (hunk 7 above)
Comment (maximburyak) [reply src/Raven.Server/Documents/Subscriptions/SubscriptionConnectionState.cs]: ok

File: src/Raven.Server/Documents/TcpHandlers/SubscriptionConnection.cs (hunk 8 above)
Comment (maximburyak) [reply src/Raven.Server/Documents/TcpHandlers/SubscriptionConnection.cs]: changed to something else

File: src/Raven.Server/Documents/TcpHandlers/SubscriptionConnection.cs (hunk 9 above)
Comment (maximburyak) [reply src/Raven.Server/Documents/TcpHandlers/SubscriptionConnection.cs:166]: Yes, if the client died we want to know it as soon as possible

Comment (ppekrol) [reply src/Raven.Server/Documents/TcpHandlers/SubscriptionConnection.cs:166]: @maximburyak I'm talking about the text here.

It will be: `Client from IP Waiting '127.0.0.1' waiting for...`

Comment (ppekrol) [reply src/Raven.Server/Documents/TcpHandlers/SubscriptionConnection.cs:166]: It should be probably `Client from IP '127.0.0.1' waiting for...`

Comment (maximburyak) [reply src/Raven.Server/Documents/TcpHandlers/SubscriptionConnection.cs:166]: Correct, too much copy pasting, will be fixed in next PR

Comment (ppekrol) [reply src/Raven.Server/Documents/TcpHandlers/SubscriptionConnection.cs:166]: Already fixed: https://github.com/ravendb/ravendb/commit/912c34ef475d94d8c11a191024ea021a105af0c1
